File: src/main/java/io/lsdconsulting/lsd/distributed/interceptor/captor/http/HttpHeaderRetriever.java
package io.lsdconsulting.lsd.distributed.interceptor.captor.http;

import feign.Request;
import feign.Response;
import io.lsdconsulting.lsd.distributed.interceptor.captor.header.Obfuscator;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpRequest;
import org.springframework.http.client.ClientHttpResponse;

import java.util.Collection;
import java.util.Map;

import static java.util.stream.Collectors.toMap;

@RequiredArgsConstructor
public class HttpHeaderRetriever {

    private final Obfuscator obfuscator;

    public Map<String, Collection<String>> retrieve(HttpRequest request) {
        return obfuscator.obfuscate(request.getHeaders().entrySet().stream().collect(toMap(Map.Entry::getKey, Map.Entry::getValue)));
    }

    public Map<String, Collection<String>> retrieve(ClientHttpResponse response) {
        return obfuscator.obfuscate(response.getHeaders().entrySet().stream().collect(toMap(Map.Entry::getKey, Map.Entry::getValue)));
    }

    public Map<String, Collection<String>> retrieve(Request request) {
        return obfuscator.obfuscate(request.headers());
    }

    public Map<String, Collection<String>> retrieve(Response response) {
        return obfuscator.obfuscate(response.headers());
    }
}

File: src/main/kotlin/io/lsdconsulting/lsd/distributed/interceptor/captor/http/HttpHeaderRetriever.kt
package io.lsdconsulting.lsd.distributed.interceptor.captor.http

import feign.Request
import feign.Response
import io.lsdconsulting.lsd.distributed.interceptor.captor.header.Obfuscator
import org.springframework.http.HttpRequest
import org.springframework.http.client.ClientHttpResponse

class HttpHeaderRetriever(
    private val obfuscator: Obfuscator
) {
    fun retrieve(request: HttpRequest): Map<String, Collection<String>> =
        obfuscator.obfuscate(request.headers.entries.associate { it.key to it.value })

    fun retrieve(response: ClientHttpResponse): Map<String, Collection<String>> =
        obfuscator.obfuscate(response.headers.entries.associate { it.key to it.value })

    fun retrieve(request: Request): Map<String, Collection<String>> =
        obfuscator.obfuscate(request.headers())

    fun retrieve(response: Response): Map<String, Collection<String>> =
        obfuscator.obfuscate(response.headers())
}
